certbot-apache/certbot_apache/interfaces.py

```diff
@@ -189,45 +194,44 @@ class DirectiveNode(ParserNode):
     It can have zero or more parameters attached to it. Because of the nature of
     single directives, it is not able to have child nodes and hence it is always
     treated as a leaf node.
-    """
 
-    @property
-    @abc.abstractmethod
-    def enabled(self):  # pragma: no cover
-        """
-        Configuration blocks may have conditional statements enabling or disabling
-        their contents. This property returns the state of this DirectiveNode.
+    If a this directive was defined on the httpd command line, the ancestor instance
+    variable for this DirectiveNode should be None, and it should be inserted to the
+    beginning of root BlockNode children sequence.
 
-        :returns: True if the DirectiveNode is parsed and enabled in the configuration.
-        :rtype: bool
-        """
+    DirectiveNode objects should have the following attributes:
 
-        raise NotImplementedError
+    # True if this DirectiveNode is enabled and False if it is inside of an
+    # inactive conditional block.
+    enabled: bool
 
-    @property
-    @abc.abstractmethod
-    def name(self):  # pragma: no cover
-        """
-        Name property contains the name of the directive.
-
-        :returns: Name of this node
-        :rtype: str
-        """
+    # Name, or key of the configuration directive
+    name: str
```

Comment:
While I do like this approach, it can make life harder in other parts of the code, as BlockNodes would also match `isinstance(DirectiveNode, ...)`. This is _mainly_ a concern regarding assertion tests, but maintaining the correct order to check the types might prove unnecessarily hard.

Reply:
I don't think it'll be that bad. I don't think it'll come up when using these objects normally because you're normally dealing with a node for specific directives and I only see two instances of `isinstance(var, DirectiveNode)` in #7338.

I'm happy to chat about this here, Mattermost, or during a 1-1, but I honestly feel pretty strongly about this. I think there's a significant benefit of code reuse and I think it results in a more accurate design. Of course a `BlockNode` is a `DirectiveNode`. It's just a special type of directive that can contain other directives.

As a potential compromise, what if we made a `BaseDirectiveNode` and `DirectiveNode` and `BlockNode` that inherit from it? `DirectiveNode` adds nothing but a distinct class name that is unrelated to BlockNodes.

Reply:
I do like the compromise option but at the same time I'm not very happy with the added complexity that could possibly be avoided. Because of this, I think I'm ok with going with the original proposal, and we'll change this later if a need arises.